### PR TITLE
Replace event cause system

### PR DESCRIPTION
**Pull request by Deamon5550**
This PR proposes changes to reduce the amount of obscurity as to the actual type of primary cause returned from an event by making Cause generic and expanding the Reason system while preserving the functionality of the recursive nature that Causes currently have.

Events which have had their reasons created for them:
- `EntityTeleportEvent` -> `TeleportReason`
- `EntityChangeHealthEvent` -> `HealthChangeReason`
- `EntityDeathEvent` -> `DamageReason`
- `PlayerDeathEvent` -> `DamageReason`
- `EntitySpawnEvent` -> `EntitySpawnReason`
- `EntityAboutToConstructEvent` -> `EntitySpawnReason`

Events which have had cause removed:
- `BlockEvent` -> The cause has been removed from this base event in favor of selectively adding it where applicable to its sub events. (Break/Place events are the ones I'm looking at).
- `EntityExpireEvent` -> The cause of an entities expiration is just that, expiration.
- `ProjectileLaunchEvent` -> The cause of a projectile launch is inferred by the included projectile source.

A builder has also been added for the creation of more complex or custom damage reasons.

__Some example reason chains for two of the currently defined reasons:__

`EntitySpawnReason` (currently used in the `EntitySpawnEvent` and `EntityAboutToConstructEvent`)
- BLOCK_SPAWNING -> `BlockReason`
- DISPENSE -> `BlockReason`
- DROPPED_ITEM -> `BlockReason`/`EntityReason`
- EXPERIENCE -> `BlockReason`/`EntityReason`
- FALLING_BLOCK -> `BlockReason`
- MOB_SPAWNER -> `BlockReason`
- PLACEMENT -> `PlayerReason`
- PROJECTILE -> `ProjectileSourceReason`
- SPAWN_EGG -> `BlockReason`/`EntityReason`
- TNT_IGNITE -> `BlockReason`

`HealthChangeReason` (currently used in the `EntityChangeHealthEvent` and `EntityDeathEvent`)
- FOOD -> `ItemReason`
- IN_FIRE -> `BlockReason`
- LAVA -> `BlockReason`
- DROWN -> `BlockReason`
- IN_WALL -> `BlockReason`
- CACTUS -> `BlockReason`
- ANVIL -> `EntityReason`
- FALLING_BLOCK -> `EntityReason`

### CODE REVIEW

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java [hunk 1]
```diff
@@ -621,17 +589,15 @@ public static EntityPickUpItemEvent createEntityPickUpItem(Game game, Entity ent
...
-     * @param cause The cause of the event, can be null
      * @param entity The entity involved in this event
      * @param block The block affected by this event
      * @param replacementBlock The block that will replace the existing block
      * @return A new instance of the event
      */
-    public static EntityPlaceBlockEvent createEntityPlaceBlock(Game game, Cause cause, Entity entity, BlockLoc block,
-            BlockSnapshot replacementBlock) {
+    public static EntityPlaceBlockEvent createEntityPlaceBlock(Game game, Entity entity, BlockLoc block,
+                                                               BlockSnapshot replacementBlock) {
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: Why did you change the indentation here?

File: src/main/java/org/spongepowered/api/event/SpongeEventFactory.java [hunk 2]
```diff
@@ -1134,18 +1113,16 @@ public static PlayerPickUpItemEvent createPlayerPickUpItem(Game game, Player pla
...
      * @param player The player involved in this event
      * @param block The block affected by this event
      * @param replacementBlock The block that will replace the existing block
      * @param direction The direction the block was placed
      * @return A new instance of the event
      */
-    public static PlayerPlaceBlockEvent createPlayerPlaceBlock(Game game, Cause cause, Player player, BlockLoc block,
-            BlockSnapshot replacementBlock, Direction direction) {
+    public static PlayerPlaceBlockEvent createPlayerPlaceBlock(Game game, Player player, BlockLoc block,
+                                                               BlockSnapshot replacementBlock, Direction direction) {
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/SpongeEventFactory.java]: Why did you change the indentation here?

File: src/main/java/org/spongepowered/api/event/cause/Cause.java [hunk 3]
```diff
@@ -25,76 +25,16 @@
...
-     *
-     * @return The reason
+     * Gets the name of this cause.
+     * 
+     * @return The name
      */
-    public Optional<Reason> getReason() {
-        return this.reason;
-    }
+    String getCauseName();
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/event/cause/Cause.java]: Any special reason to use `getCauseName()` instead of just `getName()`?

File: src/main/java/org/spongepowered/api/GameRegistry.java [hunk 4]
```diff
@@ -550,4 +559,65 @@
      */
     RecipeRegistry getRecipeRegistry();
 
+    /**
+     * Gets the {@link HealthChangeCause} with the given name.
+     * 
+     * @param name The name of the cause
+     * @return The {@link HealthChangeCause} or Optional.absent() if not found
+     */
+    Optional<HealthChangeCause> getHealthChangeChause(String name);
```
Comment (Sannholm) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: Chause -> Cause

**Comment by Zidane** [comment]
@sk89q I'll leave this to you as you wanted causes.

**Comment by Deamon5550** [comment]
I've updated the intent of the PR, please see the updated main description.

@sk89q I believe that this is now closer to your original intention for the Cause system, let me know what you think. It should have all of the flexibility and power of your original system but with slightly less obscurity by making the Cause generic.

File: src/main/java/org/spongepowered/api/event/cause/reason/DamageTypes.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * An enumeration of standard {@link DamageType}s.
+ */
+public final class DamageTypes {
+
+    public static final DamageType FIRE = null;
+    public static final DamageType MAGIC = null;
+    public static final DamageType EXPLOSIVE = null;
+    public static final DamageType PROJECTILE = null;
+
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/event/cause/reason/DamageTypes.java]: There should also be `CONTACT`, for blocks such as lava and cacti.

Additionally, there should be `CUSTOM` or `UNKNOWN` for mod or plugin caused damage.

Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/event/cause/reason/DamageTypes.java]: `UNKNOWN` isn't needed as the type is optional. `DamageReason`s can have as many types as required, or none.

I'll add the others.

File: src/main/java/org/spongepowered/api/GameRegistry.java [hunk 6]
```diff
@@ -550,4 +562,99 @@
      */
     RecipeRegistry getRecipeRegistry();
 
+    /**
+     * Gets the {@link BlockReason} with the given name.
+     * 
+     * @param name The name of the reason
+     * @return The {@link BlockReason} or Optional.absent() if not found
+     */
+    Optional<BlockReason> getBlockReason(String name);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/GameRegistry.java]: Isnt the actual block missing as param here?
Its required for getBlock inside BlockReason.

**Comment by sk89q** [comment]
I haven't gotten around to looking at everything but the removal of cause from all events means you cannot throw an event such as "block at x1, y1, z2 placed block ay x2, y2, z2" without creating your own event that no one else will listen to.

**Comment by Deamon5550** [comment]
I'm not sure I see the difference between adding an event no one will listen to or a cause no one will check. Adding arbitrary causes to events is all well and good, but it relies on the listeners of the event looking for these causes.

Preferable as I see it is having the first layer of causes defined with additional causes chained after it in an arbitrary and undefined fashion. So the `BlockPlaceEvent` would have a defined first cause one of which may be `BLOCK_SPREAD` or something, which could then have a parent cause referencing the second block to be referenced if required.

**Comment by sk89q** [comment]
Plenty of gameplay, protection and logging plugins would listen to `Cause`.

WorldGuard already [listens to a Cause in Bukkit](https://github.com/sk89q/WorldGuard/blob/master/src/main/java/com/sk89q/worldguard/bukkit/listener/RegionProtectionListener.java), which it [generates on its own](https://github.com/sk89q/WorldGuard/blob/master/src/main/java/com/sk89q/worldguard/bukkit/listener/EventAbstractionListener.java).

The primary purpose of Cause is **not** to track parent causes. Cause exits *because it's unfeasible to cause parent causes*.

For example, a [CraftBook gate triggered by Redstone](https://www.youtube.com/watch?v=JEHbP0TuaFk#t=85) may be ultimtaely caused by a player, but because it's not reasonable to trace it to the player, Cause allows an event to be fired with the cause being **the block**.

Saplings that grow into trees may ultimately have been placed by a player, but because that's not possible to trace, [the Cause becomes the sapling itself](https://github.com/sk89q/WorldGuard/blob/master/src/main/java/com/sk89q/worldguard/bukkit/listener/EventAbstractionListener.java#L175).

**Comment by Deamon5550** [comment]
Yes, I think we agree on what a cause is or isn't. Where I think we differ is the type of information available in a cause.

If we take example of a player taking some damage from a source (Our EntityChangeHealthEvent, bukkit's EntityDamageEvent, worldguard's DamageEntityEvent) there are two (that I'll present here anyway) options for what information to include in the cause.

**Option A** (the world guard way): the cause includes only the damager, either block or entity.

**Option B** (my current proposition): the cause includes first more general cause information followed then by additional information, such as a damager if one exists.

If we narrow our example to be an entity standing in a fire block. Option A gives us a cause which contains only the fire block and leaves the listener to infer the reason of the damage being caused by that block. Whereas with option B our cause would have the structure (IN_FIRE, BlockLoc) so the reason for the damage is explicit.

Option B provides overall more information and leaves some cases as being less ambigious. For example with Option A it is impossible to detect the difference between damage from falling or damage from starving.

If I have misinterpreted the usage and extent of causes in worldguard please correct me.

**Comment by sk89q** [comment]
There is option C, which is the current Sponge way (although not well developed), which is an extension of option A with general cause information.

In your example of an entity of a fire block, under option C:

* Cause would refer to the fire block
* Cause.getReason() would refer to IN_FIRE

C, which is currently what is in Sponge, is a superset of both other options.

**Comment by Deamon5550** [comment]
However, it seems to me that 'cause' and 'reason' are essentially the same thing. One issue I have with option C is that the cause is non-optional while the reason is optional. With the example of falling damage the reason would be FALLING and the cause would be...null?

So I think it is applicable to a wider range of uses to have cause and reason as the same thing and chained in the fashion of option B, and we still aren't losing any information with option B over C although I think I'll streamline the objects a bit more to make them more easily usable (so its not cause.getReason().getEntity() the reason can just be the entity or something like that, thoughts on this?).

*(I'd argue that the optional on cause and reason could be reversed but in my opinion this still doesn't solve the issue of being applicable to the widest array of use cases)*

**Comment by sk89q** [comment]
The cause could be the world. Personally I feel that the number of cases that are caused by a block or entity vastly outnumber the cases that don't. Pretty much every event that we have involves a player, entity, or block.

The reason I am generally for cause first is because this is the easy information to provide for most mods and plugins that modify the world, which is important because we should expect the absolute worse from developers. Its basis is on objects that already exist in Minecraft (entity, block, etc.).

Reasons are a Sponge invention, and their proper use means that a custom Reason may have to be provided, so in most cases, we can assume (assuming the worst) that the developer would just pass UNKNOWN and would forget (or more likely, not know) to specify the causing block or entity. 

But as I mentioned, my main issue is not "reason first" or "cause first" -- so everything above is unimportant in the grand scheme in my opinion. My issue is that the PR doesn't implement A, B, or C -- there is no reason or cause **at all** on the events. It implements Bukkit's way.

BlockChangeEvent is completely useless after these changes. There is no way at all to fire an event (without creating a new one that no one will listen to) to say that a block has changed another block.

**Comment by Zidane** [comment]
@SpongePowered/developers Feedback please.

@Deamon5550 I think Cause needs to remain an integral part of Events. Find a way to happily marriage Cause and Reason.

**Comment by Deamon5550** [comment]
Yeah, its not my intention to remove the concept of a cause. I've got a fair idea of where I want to run with this, but I'll need to rewrite it.

Closing this for now.
